stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+ */
+template <typename... T_expressions>
+class eigen_expressions_ {
+ public:
+  /**
+   * Constructor.
+   * @param exprs expresions that will be calculated in same loop.
+   * @throw invalid_argument expressions have different sizes (for row major
+   * expressions rows and columns are swapped for the check)
+   */
+  explicit eigen_expressions_(T_expressions&&... exprs)
+      : exprs_(std::forward<T_expressions>(exprs)...) {
+    index_apply<sizeof...(T_expressions) - 1>([&](auto... Is) {
+      constexpr auto first_flags = Eigen::internal::traits<std::decay_t<
+          std::tuple_element_t<0, std::tuple<T_expressions...>>>>::Flags;
+      static_cast<void>(std::initializer_list<int>{
+          ((((Eigen::internal::traits<std::decay_t<std::tuple_element_t<
+                  Is + 1, std::tuple<T_expressions...>>>>::Flags
+              ^ first_flags)
+             & Eigen::RowMajorBit)
+                ? check_matching_dims("eigen_expressions_.eigen_expressions_",
+                                      "first expression", std::get<0>(exprs_),
+                                      "transposed expression",
+                                      std::get<Is + 1>(exprs_).transpose())
+                : check_matching_dims("eigen_expressions_.eigen_expressions_",
+                                      "first expression", std::get<0>(exprs_),
+                                      "expression", std::get<Is + 1>(exprs_))),
+           0)...});
+    });
+  }
+
+ private:
+  std::tuple<T_expressions...> exprs_;
+  template <typename... T_results>
+  friend class eigen_results_;
+};
+
+/**
+ * Deduces types for constructing \c expressions_ object.
+ * @tparam T_expressions types of exprs
+ * @param exprs exprs that will be used in same kernel.
+ */
+template <typename... T_expressions,
+          require_all_eigen_t<T_expressions...>* = nullptr>
+eigen_expressions_<T_expressions...> eigen_expressions(
+    T_expressions&&... exprs) {
+  return eigen_expressions_<T_expressions...>(
+      std::forward<T_expressions>(exprs)...);
+}
+
+/**
+ * Represents results that will be calculated in same loop.
+ * @tparam T_results types of results
+ */
+template <typename... T_results>
+class eigen_results_ {
+  std::tuple<T_results...> results_;
+
+  /**
+   * Assign expressions to results using linear indexing.
+   * @tparam Linear whether to use linear indexing
+   * @tparam T_expressions types of expressions
+   * @param expressions expressions to assign
+   */
+  template <bool Linear, typename... T_expressions,
+            std::enable_if_t<Linear>* = nullptr>
+  void assign(const eigen_expressions_<T_expressions...>& expressions) {
+    for (size_t i = 0; i < std::get<0>(expressions.exprs_).size(); i++) {
+      index_apply<sizeof...(T_results)>([&](auto... Is) {
+        static_cast<void>(std::initializer_list<int>{
+            (std::get<Is>(results_).coeffRef(i)
+             = std::get<Is>(expressions.exprs_).coeff(i),
+             0)...});
```

Comment:
Should we just use the foreach thing I wrote a while ago for this? That at least let's you get rid of the static_cast etc. here

https://gist.github.com/SteveBronder/410f7127f8be9dced2b1915765fd8644

Reply:
We could just as well, although the difference is not that big. Also I think that foreach is not already in the repo, right?